Admin new problem page, Submissions using DB, Cipher List

andersonjackc opened the pull request:
While logged in as an admin, you can create a new problem with a UI, that will appear in the problems list.

When answering a question, an AnswerSubmission is made to the DB which updates Problems Attempts/Completions field. Logic was also implemented relating to who can see the submit fields for a problem and when.

A cipher list nav bar and page was added, and while the link works, it redirects you to the Caesar page for all name attributes, but this is fine as we do not have auto-generating cipher pages yet.

Some session information could be implemented for some of the pages, but that can be discussed.

andersonjackc committed:
I guess .sln was never added to GitIgnore?

andersonjackc committed:
AddProblem Controller

andersonjackc committed:
Create AddProblemPage.cshtml

andersonjackc committed:
Added AddProblems to Nav Bar if you are an Admin

andersonjackc committed:
Only Users can submit Answers

andersonjackc committed:
Added ProblemID field

This is also in the DB

andersonjackc committed:
Added some extra logic to the problems

andersonjackc committed:
Several Queries

Update Problem, InsertAnswerSubmission, GetProblemCorrectValueByUserAndProblemID

andersonjackc committed:
Added some Problem Logic

Problems now have a submission, and the attempts/completions are updated

andersonjackc committed:
Working on Cipher Nav Bar

andersonjackc committed:
Fixed a comment

andersonjackc committed:
Edited CipherController for Generic Page generation

andersonjackc committed:
Edited Cipher.cshtml for generic page generation

andersonjackc committed:
Updated Startup.cs for page generation

andersonjackc committed:
Update ProblemController.cs

jgross11 commented on .gitignore:
```diff
@@ -350,3 +350,4 @@ MigrationBackup/
 .ionide/
 *.sln
 Project EFT.sln
```
already in gitignore on line 352, but it was necessary to push sln changes when adding testing portion of project, so not completely stagnant

jgross11 commented on Project EFT/Controllers/ProblemController.cs:
```diff
@@ -74,6 +74,33 @@ public IActionResult CheckAnswer()
+            }
+            else
+            {
+                DBConnector.InsertNewProblem(problem);
```
what happens if `InsertNewProblem` fails?

jgross11 replied:
In reference to the next line, which got cut off in the MLC:
`ViewData["message"] = "Your new problem has been added to the list of problems!";`

andersonjackc replied:
Now it checks what is returned, true/false, and sends a message accordingly.

jgross11 commented on Project EFT/Views/Shared/AddProblemPage.cshtml:
```diff
@@ -0,0 +1,33 @@
+    <!-- submission form-->
+    <form action="~/Problem/AddProblem" method="post">
+        Enter Problem Title: <br />
+        <input type="text" id="title" name="title" /><br />
+        Enter Problem Question: <br />
+        <textarea name="question" id="question" cols="50" rows="4"> </textarea> <br />
+        Enter Problem Answer: <br />
+        <textarea name="answer" id="answer" cols="50" rows="4"> </textarea> <br />
```
will eventually have to enforce character limits

jgross11 commented on Project EFT/Views/Shared/AddProblemPage.cshtml:
```diff
@@ -0,0 +1,33 @@
+        <div></div>
+    </form>
+
+    @if (ViewData["message"] == null)
+    {
+
+    }
+    else
+    {
+        <div>@ViewData["message"]</div>
+    }
```
if the value at that key is null, the div is empty -> no logic needed

andersonjackc replied:
Updated this, and I used this same technique in a different place as well.  Fixed both.

jgross11 commented on Project EFT/Views/Shared/Problem.cshtml:
```diff
@@ -19,6 +19,10 @@
             <div></div>
         </form>
     }
+    else if (session.ContainsKey("userInfo") && !((bool)ViewData["showProblem"]))
+    {
+        <br /><br /><div>Congratulations, this problem has been completed!</div>
+    }
```
if contains key
---- if value in map is true
---- else
else

jgross11 replied:
In reference to this and the preceding code block, which got cut off in the MLC:
`@if (session.ContainsKey("userInfo") && (bool)ViewData["showProblem"])`
[this code snippet]
`else{ [not logged in as a user case] }`

andersonjackc replied:
Updated

jgross11 commented on Project EFT/Database/DBConnector.cs:
```diff
@@ -239,6 +286,22 @@ public static Problem GetProblemByID(int ID)
+            MySqlCommand command = MakeCommand("SELECT AnswerSubmissions_IsCorrect FROM AnswerSubmissions WHERE User_ID = @user_id AND AnswerSubmissions_ProblemID = @problem_id AND AnswerSubmissions_IsCorrect = @correct");
+            command.Parameters.AddWithValue("@user_id", userID);
+            command.Parameters.AddWithValue("@problem_id", problemID);
+            command.Parameters.AddWithValue("@correct", true);
```
can this be part of the statement, since it always takes the same value?

andersonjackc replied:
Yes it can.

jgross11 commented on Project EFT/Database/DBConnector.cs:
```diff
@@ -153,6 +154,20 @@ public static int UpdatePassword(User user, string newPassword)
             return result == 1 ? CREDENTIAL_CHANGE_SUCCESS : DB_FAILURE;
         }
 
+        public static int UpdateProblem(Problem problem)
+        {
+            MySqlCommand command = MakeCommand("UPDATE Problems SET Problem_Title = @title, Problem_Answer = @answer, Problem_Attempts = @attempts, Problem_Completions = @completions WHERE Problem_Number = @id");
+            command.Parameters.AddWithValue("@title", problem.Title);
+            command.Parameters.AddWithValue("@answer", problem.Answer);
+            command.Parameters.AddWithValue("@attempts", problem.Attempts);
+            command.Parameters.AddWithValue("@completions", problem.Completions);
+            command.Parameters.AddWithValue("@id", problem.ProblemNumber);
+            command.Prepare();
+            int result = command.ExecuteNonQuery();
+
+            return result;
+        }
```
presumably, this will only be used for updating attempts / completions, so is it necessary to also 'update' the title, answer, etc.?

andersonjackc replied:
I was considering long term where we may want a version of the admin page that updates a problem, because someone mistypes the answer to the cipher, or a bad admin adds a crap problem, something like that.  Rather than just deleting it, we would just update that problem.  I can edit it if you feel its necessary, but I just felt it might be better to have the more general version.

jgross11 commented on Project EFT/Database/DBConnector.cs:
```diff
@@ -208,6 +223,38 @@ public static bool InsertNewAdminSubmission(Submission submission)
+            Problem problem = GetProblemByID(submission.ProblemId);
+
+            //updates the correct field in the problem based on attempts vs. completions
+            if (submission.IsCorrect)
+            {
+                problem.Attempts++;
+                problem.Completions++;
+            }
+            else
+            {
+                problem.Attempts++;
+            }
```
attempts incremented in both cases

andersonjackc replied:
Updated

andersonjackc committed:
Fixed null checks, condensed logic statements

andersonjackc committed:
Updated some logic layout in Problem.cshtml

andersonjackc committed:
Added a value directly to the Query

andersonjackc committed:
Attempts is always incremented

andersonjackc committed:
Added a check if the Insert fails.

andersonjackc committed:
Encrypt is fixed, decrypt...im working on

andersonjackc committed:
Fixed a decrypt issue

% =/= arithmetic modulus

andersonjackc committed:
Encrypt/Decrypt Update

Encryption/Decryption now works with shifts less than zero or greater than the alphabet size

andersonjackc committed:
Update Cipher.cshtml
